logic/strategy/src/one_hop_connection.rs

```diff
@@ -0,0 +1,358 @@
+            if !other_outgoing_channels_from_destination_peer.is_empty() {
+                // From all the outgoing channels from the destination peer that points to a relay node outside of the destination peer list,
+                // get the subset of relay nodes that that we have an outgoing channel to, and has a good network quality.
+                let relay_nodes = other_outgoing_channels_from_destination_peer
+                    .iter()
+                    .filter_map(|channel| {
+                        let relay_node = channel.destination;
+                        // Check if we have an outgoing channel to the relay node
+                        if outgoing_channels
+                            .clone()
+                            .any(|outgoing_c| outgoing_c.destination == relay_node)
+                        {
+                            let peers_with_quality_cloned = peers_with_quality.clone();
+                            let relay_node_quality = {
+                                let quality = peers_with_quality_cloned.get(&relay_node);
+                                quality.unwrap_or(&0_f64)
+                            };
+                            if *relay_node_quality >= self.cfg.network_quality_threshold {
+                                Some((relay_node, channel.balance))
+                            } else {
+                                None
+                            }
+                        } else {
+                            None
+                        }
+                    })
+                    .collect::<Vec<(Address, Balance)>>();
+                if relay_nodes.is_empty() {
+                    should_find_new_relay_node = true;
+                } else {
+                    should_find_new_relay_node = false;
+
+                    // Among those relay nodes with good network quality, top up channels to relay nodes with low balance.
+                    for (relay_node, outgoing_channel_to_relay_node_balance) in relay_nodes {
+                        // check if the channel needs to be topped up
+                        if outgoing_channel_to_relay_node_balance.lt(&self.cfg.min_stake_threshold) {
+                            if !tick_decision.will_address_be_opened(&relay_node) {
+                                tick_decision.add_to_open(relay_node, self.cfg.funding_amount);
+                                remaining_balance = remaining_balance.sub(&self.cfg.funding_amount);
+                                debug!(%relay_node, "relay node will be topped up");
+                            } else {
+                                debug!(%relay_node, "relay node is already promoted for topping up");
+                            }
+                        } else {
+                            // Keep existing relay nodes alive
+                            debug!(%relay_node, "relay node is alive and has enough balance");
+                        }
+                    }
```

Comment:
This piece of code is potentially prone to opening vulnerabilities for adversarial actions with scaling effects. I.e. for every node in the network that runs this strategy towards a relayer R, the relayer R can by closing a single channel to target T and affect the expenditures of all channels of N * M nodes that are funded between N and R, i.e. a 1:M draining scale coefficient pretty predictably.

Reply:
Handled by the comment.